lib/deepcrawl.js

```diff
@@ -0,0 +1,137 @@
+'use strict';
+const BPromise = require('bluebird'),
+  schema = require('../schemas/v2.0.0'),
+  needle = BPromise.promisifyAll(require('needle')),
+  humps = require('humps'),
+  _ = require('lodash');
+
+class DeepCrawl {
+  constructor(cfg) {
+    cfg = cfg || {};
+
+    this.apiId = cfg.apiId;
+    this.apiKey = cfg.apiKey;
+    this.getSessionToken = cfg.getSessionToken;
+    this.accountId = cfg.accountId;
+
+    this.schema = cfg.schema || schema;
+    this.version = this.schema.version;
+    this.baseUrl = this.schema.baseUrl;
+  }
+
+  createEndpoints (schema) {
+    const API = {};
+    for (const resource in schema.resources) {
+      API[resource] = {};
+    }
+    return API;
+  }
+
+  generateRequest (url, options, method) {
+    // first, get the sessionToken and generate the request options and headers
+    return BPromise.resolve(options.sessionToken || this.getSessionToken())
+      .then((sessionToken) => {
+        const requestOpts = {
+          json: true,
+          headers: {
+            'X-Auth-Token': sessionToken
+          }
+        };
+
+        if (method === 'get') {
+          return needle.getAsync(url, requestOpts)
+            .then(this.handleResponse);
+        }
+
+        // convert all camelCased options to _ for deepcrawl
+        const data = humps.decamelizeKeys(options);
+        return needle[`${method}Async`](url, data, requestOpts)
+          .then(this.handleResponse);
+      });
+  }
+
+  generateMethod (requiredFields, url, method) {
+    return (options) => {
```

Comment:
Can you add some inline comments here that returning this function as a closure is the key to constructing it as a dynamic method call so that future knuckleheads like me won't also be confused.